File: Cplusplus/4Sum.cpp
/*
Leetcode : 4Sum.cpp

Given an array S of n integers, are there elements a, b, c,

and d in S such that a + b + c + d = target? Find all unique

quadruplets in the array which gives the sum of target.

Note:
Elements in a quadruplet (a,b,c,d) must be in non-descending

order. (ie, a ≤ b ≤ c ≤ d)
The solution set must not contain duplicate quadruplets.
    For example, given array S = {1 0 -1 0 -2 2}, and target = 0.

    A solution set is:
    (-1,  0, 0, 1)
    (-2, -1, 1, 2)
    (-2,  0, 0, 2)

*/
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<vector<int> > fourSum(vector<int> &num, int target) {
        sort(num.begin() , num.end());
        vector<vector<int> > result;
        if(num.size()<4)
            return result;
        for(int j = 0 ; j < num.size() - 3 ; j++)
        {
            if(j!=0 && num[j] == num[j-1])
                continue;
            for(int i = j+1 ; i < num.size() - 2 ; i++)
            {
                if(i!= j+1  && num[i] == num[i-1])
                    continue;
                int left = i+1;
                int right = num.size() - 1;

                while(left < right)
                {
                    if(left > i+1 && num[left] == num[left-1])
                    {
                        left++;
                        continue;
                    }
                    if(num[left] + num[right] + num[i] == target - num[j])
                    {

                        vector<int> tmp;
                        tmp.push_back(num[j]);
                        tmp.push_back(num[i]);
                        tmp.push_back(num[left]);
                        tmp.push_back(num[right]);
                        result.push_back(tmp);


                        left ++;
                    }
                    else if (num[left] + num[right] + num[i] < target - num[j])
                        left++;
                    else
                        right--;
                }
            }
        }
        return result;
    }
};

int main()
{
    Solution test = Solution();
    int arr[6] = {1, 0, -1, 0, -2 ,2};
    vector<int> num(arr , arr+6) ;
    int target = 0;
    vector<vector<int> > result = test.fourSum(num , target);
    for(int i = 0 ; i < result.size() ; i++)
    {
        for(int j = 0 ; j < result[0].size() ;j++)
            cout<<result[i][j]<<" ";
        cout<<endl;
    }
}

File: python/4Sum.py
'''
Created on 2014-8-18

Leetcode : 4Sum

Given an array S of n integers, are there elements a, b, c, and d in S such that

 a + b + c + d = target? Find all unique quadruplets in the array which gives the sum of target.

Note:
Elements in a quadruplet (a,b,c,d) must be in non-descending order. (ie, a <= b <= c <= d)
The solution set must not contain duplicate quadruplets.
    For example, given array S = {1 0 -1 0 -2 2}, and target = 0.

    A solution set is:
    (-1,  0, 0, 1)
    (-2, -1, 1, 2)
    (-2,  0, 0, 2)
'''
class Solution:
    # @return a list of lists of length 4, [[val1,val2,val3,val4]]
    def fourSum(self, num, target):
        num.sort()
        dict = {}
        for i in range(len(num)):
            for j in range(i+1 , len(num)):
                if dict.has_key(num[i]+num[j]):
                    dict[num[i]+num[j]].append([i,j,num[i] , num[j]])
                else:
                    dict[num[i]+num[j]] = [ [i,j, num[i] , num[j] ] ]
        
        r = []
       
        keys = dict.keys()
        keys.sort()
        for key in keys:
            if target - key >= key:
                if dict.has_key(target - key):
                    for i in range(len(dict[key])):
                        for j in range(len(dict[target - key])):
                            index = dict[key][i][:2] + dict[target - key][j][:2]
                            #print index , len(set(index))
                            if len(set(index)) == 4:
                                newCom = dict[key][i][2:] + dict[target - key][j][2:]
                                newCom.sort()
                                if newCom not in r:
                                    r.append(newCom)
                
        return r
            
        
        
        
        
        
        

test = Solution()
num = [1 ,0, -1, 0 ,-2 ,2]
target = 0
#num = [-491,-468,-450,-415,-414,-374,-357,-333,-305,-292,-274,-272,-271,-258,-241,-240,-227,-217,-196,-184,-161,-120,-116,-110,-110,-98,-92,-47,-27,-10,-8,-7,-4,-1,19,30,53,62,64,65,137,138,145,160,178,179,209,221,243,270,279,283,290,291,305,308,322,345,354,357,365,366,368,371,376,381,381,394,400,406,429,433,445,446,449,470,471,472]
#target = 2873
print test.fourSum(num, target)
